### ISSUE TITLE
Unable to verify proofs using MerkleProof.sol

### ISSUE THREAD

**Issue by gianmarcotoso** [OpenZeppelin/merkle-tree#5]
I have a list of addresses from which I generate a Merkle Tree, and I set the tree's root on the contract. I want a function on my contract to allow only addresses that are part of the tree to run it. 

I have written a test that generates a tree from 3 addresses, and after setting the root on the contract attempts to call said function using a valid address and an invalid address:

```
const MerkleTest = artifacts.require('MerkleTest')
const { StandardMerkleTree } = require('@openzeppelin/merkle-tree')

contract('MerkleTest', function (accounts) {
	it('should allow an address within the merkle root to do stuff', async function () {
		const leaves = accounts.slice(1, 3).map((a) => [a])
		const tree = StandardMerkleTree.of(leaves, ['address'])

		const instance = await MerkleTest.deployed()
		await instance.setRoot(tree.root, { from: accounts[0] })

		const proof = tree.getProof(leaves[0])

		try {
			const ok = await instance.doStuff(proof, { from: accounts[1] })
			assert.isTrue(ok)
		} catch {
			assert.fail('should not have thrown')
		}
	})

	it('should not allow an address outside the merkle root to do stuff', async function () {
		const leaves = accounts.slice(1, 3).map((a) => [a])
		const tree = StandardMerkleTree.of(leaves, ['address'])

		const instance = await MerkleTest.deployed()
		await instance.setRoot(tree.root, { from: accounts[0] })

		const proof = tree.getProof(leaves[0])

		try {
                        // Any proof will be wrong here, because we're using the wrong account
			await instance.doStuff(proof, { from: accounts[5] })
			assert.fail('should have thrown')
		} catch (error) {
			assert.isTrue(error.message.includes('Not verified'))
		}
	})
})
```

This is the smart contract:

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
import "@openzeppelin/contracts/access/Ownable.sol";

contract MerkleTest is Ownable {
    bytes32 private root;

    function setRoot(bytes32 _root) public onlyOwner {
        root = _root;
    }

    function doStuff(bytes32[] memory proof) public view returns (bool) {
        bool isVerified = MerkleProof.verify(
            proof,
            root,
            keccak256(abi.encodePacked(msg.sender))
        );

        require(isVerified, "Not verified");

        return true;
    }
}
```

I expect that calling the `doStuff` function providing the proof for the address corresponding to `msg.sender` to result in the call to `MerkleProof.verify` to succeed, but it doesn't. The same thing works correctly using the `merkletreejs` library, and since this library does seem to work correctly I assume this is a misunderstanding on my part.

I have also tried to verify using the `bytes32(abi.encodePacked(msg.sender))` instead of `keccak256(abi.encodePacked(msg.sender))`, to no avail.

How do I correctly verify a proof? Documentation seems to be lacking on this, should the docs be improved to be more clear?

**Comment by frangio** [comment]
We're missing a snippet for how to use `MerkleProof`, but the readme does state how you have to compute the leaf. Take a look at the Solidity snippet in [this section](https://github.com/OpenZeppelin/merkle-tree#standard-merkle-trees):

```solidity
bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(addr, amount))));
```

So where you're using `keccak256(abi.encodePacked(msg.sender))` you should replace with `keccak256(bytes.concat(keccak256(abi.encode(amount))))`.

Will add docs showing how to use `MerkleProof.verify`. Thanks for bringing this up.

**Comment by gianmarcotoso** [comment]
Thank you, this indeed works, I had completely skipped over that line. This however leaves me with a doubt: contracts are immutable, so it's not possible to move from `merkletreejs` to this library without publishing a new contract, is that right?

Does the design you have chosen allow to implement a "fallback"/"legacy" mode that allows to use this library with already published contracts the same approach I have used in the issue (`keccak256(abi.encodePacked(msg.sender))`)? If so, would there be any drawback or caveat to adding it or using it?

**Comment by frangio** [comment]
> contracts are immutable, so it's not possible to move from `merkletreejs` to this library without publishing a new contract, is that right?

Yes this is true. It was not our goal to enable already deployed merkle trees or merkle verifiers to migrate over to this library. If merkletreejs has been working fine there is no strong reason to migrate, but we will recommend this library for new projects.

> Does the design you have chosen allow to implement a "fallback"/"legacy" mode

Yes, this is probably relatively simple to add, the library was designed to support adding some variations in how the tree is formed. We wanted a highly focused initial release, and we plan to keep the scope of this project fairly small so it is low maintenance. Currently no plans to support this legacy mode but we will consider it if a good number of users ask for it!

> This would make a more flexible approach potentially vulnerable, is that right?

This is right.

> Maybe instead of more options a line in the README saying something along the lines of "if you need more flexibility go with `merkletreejs`" could suffice.

We would actually still recommend sticking to this library. If users need more flexibility we expect to hear user requests and will be considering them as they come.

**Comment by frangio** [comment]
We've added a Solidity verification example in the readme. See [Validating a Proof in Solidity](https://github.com/OpenZeppelin/merkle-tree#validating-a-proof-in-solidity). (https://github.com/OpenZeppelin/merkle-tree/pull/6)

**Comment by gianmarcotoso** [comment]
Thank you! 

I think there's a missing `require` here:

```
        bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(addr, amount))));
        // (3)
        bool verified = MerkleProof.verify(proof, root, leaf), "Invalid proof");
        // (4)
        // ...
```

**Comment by frangio** [comment]
Hah, funny you mention that. We debated between a `require` and a boolean variable to show that you could also react to an invalid proof. I do think generally the code will want to reject invalid proofs entirely, so a `require` might be clearer. @ernestognw What do you think?

**Comment by gianmarcotoso** [comment]
I was thrown off by the trailing `, "Invalid proof");`, which as it stands would be a syntax error, and I didn't notice that you also are saving the result in a boolean value.

Personally I would go with a `require` because I'd mostly use Merkle Proof verification within a function that evolves the state, and a hard fail with a rollback seems to be more appropriate (and more safe?). You can also ask your node to do a dry run before sending the transaction, and if the dry run fails you can avoid sending a failing transaction and wasting GAS, that's also a nice plus.

**Comment by ernestognw** [comment]
> I was thrown off by the trailing `, "Invalid proof");`, which as it stands would be a syntax error, and I didn't notice that you also are saving the result in a boolean value.
> 
> Personally I would go with a `require` because I'd mostly use Merkle Proof verification within a function that evolves the state, and a hard fail with a rollback seems to be more appropriate (and more safe?). You can also ask your node to do a dry run before sending the transaction, and if the dry run fails you can avoid sending a failing transaction and wasting GAS, that's also a nice plus.

Actually, this is a change that I pushed for but I recognize @frangio has more context and experience than me.
What's wrong here is the trailing `, "Invalid proof")`, so thank you so much for noticing that.

Now that we got specific feedback about this I think I can open a PR getting back to the `require` behavior.
